index.js

```diff
@@ -416,13 +416,13 @@ export function StateT(M) {
   //. ```
   StateT.prototype['fantasy-land/ap'] = function(mf) {
     const mx = this;
-    return new StateT (state => {
-      const get = StateT.evalState (state); // Monad {st, val}
-      return Z.map (
-        value => ({state: state, value: value}),
-        Z.ap (get (mf), get (mx))
-      );
-    });
+    return new StateT (state1 => Z.chain (
+      ({state: state2, value: f}) => Z.map (
+        ({state, value: x}) => ({state, value: f (x)}),
+        mx.run (state2)
+      ),
+      mf.run (state1)
+    ));
```

Comment:
Kind of freaky how, because we need `state2` in order to evaluate `mx`, we can no longer use `ap` on `mx.run()` and `mf.run()` directly. This means that we've had to reimplement `ap` for `m` in terms of `map` and `chain`, meaning that any optimizations `m` might've had in `ap`, these will be lost in the monad transformation. For example, a Monad with a concurrent implementation of `ap` will become sequential after being transformed with `StateT`.